### ISSUE TITLE
bug: no clear way of changing default nvim-dap configuration from extras.lang.clangd

### ISSUE THREAD

**Issue by ifonajs** [LazyVim/LazyVim#2567]
### Did you check docs and existing issues?

- [X] I have read all the LazyVim docs
- [X] I have searched the existing issues of LazyVim
- [X] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

NVIM v0.9.5

### Operating system/version

PopOS 22.04

### Describe the bug

I want to tweak extras.lang.clangd and change the DAP specification. However it's not clear how to do so as `opts` defined in the default config is a function that doesn't return or modify a table. This is contrary to what is stated in [documentation for opts](https://github.com/folke/lazy.nvim?tab=readme-ov-file#-plugin-spec). The same is with extras.lang.typescript.

### Steps To Reproduce

1. Try to change nvim-dap config for c/cpp while having extras.lang.clangd enabled has not effect.

### Expected Behavior

Default nvim-dap configuration is changed by defining another function.

**Comment by ifonajs** [comment]
Found the root cause of my problems. I was installing extras with a file defined in ~/.config/nvim/lua/plugins/extras.lua. When I moved importing to ~/.config/nvim/lua/config/lazy.lua, now I'm able to change configuration of nvim-dap.

**Comment by dpetka2001** [comment]
It mentions it in the docs for each extra that it should be imported either via `:LazyExtras` or `/lua/config/lazy.lua` manually.
> You can enable the extra with the :LazyExtras command. Plugins marked as optional will only be configured if they are installed.
Alternatively, you can add it to your lazy.nvim imports
```lua
-- lua/config/lazy.lua
require("lazy").setup({
  spec = {
    { "LazyVim/LazyVim", import = "lazyvim.plugins" },
    { import = "lazyvim.plugins.extras.lang.clangd" },
    { import = "plugins" },
  },
})
```
Could you kindly close the issue if your problem is resolved?

**Comment by ifonajs** [comment]
The example.lua suggests otherwise and it seems it only doesn't work if `opts` is a `function()` (which is also not something you should assign to `opts` according to documentation). If you think this is a user error, I will close the issue.

**Comment by dpetka2001** [comment]
What do you mean the `example.lua` suggests otherwise? I specifically quoted what is mentioned in the [clangd docs](https://www.lazyvim.org/extras/lang/clangd) itself, where it mentions how you can enable this Extra. Where did you see that `opts` can't be a function? It can be a function that returns a table. From `lazy.nvim` [docs](https://github.com/folke/lazy.nvim#-plugin-spec)
> opts should be a table (will be merged with parent specs), return a table (replaces parent specs) or should change a table. The only difference here is that the table is created and then can be modified via the required `dap` module, because that's how the `nvim-dap` plugin works.

**Comment by ifonajs** [comment]
https://www.lazyvim.org/configuration/examples has examples of importing extras in `plugins/` directory. I also wanted to have it in `plugin/` folder to have all of my modifications in one place but it's not a big deal.

> Where did you see that opts can't be a function?
More precisely, function that doesn't take arguments and doesn't return anything. But DAP is a special in this case as you described. From the same `lazy.nvim` docs I could also assume `function()` is not supported at all:
> opts |  table or fun(LazyPlugin, opts:table)

From my point of view, the fact that  importing extras somewhere in `plugins/` folder works in 50% of the cases (when `opts` is a table or configured with `function(_, opts)`), means there's area for improvement. Things should work in all cases or shouldn't work at all.

**Comment by dpetka2001** [comment]
The difference that you notice when you import the Extras in your `/lua/plugins/` folder is because `lazy.nvim` parses files in alphabetical order. In your previous example you had the import in `/lua/plugins/extras.lua`, but you had your custom spec of `nvim-dap` in `/lua/plugins/clangd-extra.lua`. So, the `clangd-extra.lua` will get parsed first by lazy and then the `extras.lua` file. In this case the latter overrides the former. You can check that with the following config
```lua
  {
    "mfussenegger/nvim-dap",
    opts = function()
      local dap = require("dap")
      dap.configurations.c[1].name = "test"
    end,
  },
  { import = "lazyvim.plugins.extras.lang.clangd" },
```
It won't be like you want it to. Now if you put it like 
```lua
  { import = "lazyvim.plugins.extras.lang.clangd" },
  {
    "mfussenegger/nvim-dap",
    opts = function()
      local dap = require("dap")
      dap.configurations.c[1].name = "test"
    end,
  },
```
It will work as expected. This is why in the Extras documentation it says to put the import in your `/lua/config/lazy.lua`, because that file and everything in it will get parsed before the files in `/lua/plugins/` directory.

The safest way is to enable the Extras you want via the command `LazyExtras`, because that takes into consideration the correct ordering of parsing files.

Anyway, this is not a LazyVim bug. If you'd like you could make a PR in the [docs repo](https://github.com/LazyVim/lazyvim.github.io) or create an issue with `enhancement` label to clarify this in the LazyVim website.

**Comment by ifonajs** [comment]
What do you think about commenting out [this](https://github.com/LazyVim/starter/blob/main/lua/plugins/example.lua#L136) or putting a warning in the comments not to include Extras like that? This is what lead me to the wrong assumptions in the first place.

**Comment by dpetka2001** [comment]
Yes, I'm in favor of this because it seems to mislead new users. But I'm just a user like you with no access to this project. That's why I suggested for you to create an issue with `enhancement` label or create a PR yourself that comments out or even deletes the current section that you're referring to.

**Comment by dpetka2001** [comment]
@ifonajs Did you reopen this issue by mistake or is there any particular problem you keep having?
